### PR TITLE
Refactor/re-organising MPI related code

### PR DESCRIPTION
**Pull request by pramodk**
**Description**

  - All MPI related code is now under coreneuron/mpi
  - Separate cmake target for mpi related code and
    include MPI headers for only these files
  - Remove unnecessary headers (and MPI include)
  - Spike output function is factored into a separate file

Preparation for #600

**How to test this?**

No new options/instructions.

**Use certain branches for the SimulationStack CI**

CI_BRANCHES:NEURON_BRANCH=master,

### CODE REVIEW

File: coreneuron/CMakeLists.txt [hunk 1]
```diff
@@ -170,7 +179,7 @@ target_compile_options(coreneuron
                        PRIVATE $<$<COMPILE_LANGUAGE:CXX>:${CORENEURON_CXX_WARNING_SUPPRESSIONS}>)
 
 # need to have _kinderiv.h for mod2c generated files and nrnivmodl-core and nmodl building
-add_dependencies(coreneuron kin_deriv_header nrnivmodl-core)
+add_dependencies(coreneuron kin_deriv_header corenrn_mpi nrnivmodl-core)
```
Comment (olupton) [inline coreneuron/CMakeLists.txt:183]: Is this needed? Looks redundant with `$<TARGET_OBJECTS:corenrn_mpi>` above.

**Review by olupton — COMMENTED** [review]
LGTM, one minor question.

I guess you need to rebase on top of https://github.com/BlueBrain/CoreNeuron/pull/623 for the CI?

File: coreneuron/CMakeLists.txt (hunk 1 above)
Comment (pramodk) [reply coreneuron/CMakeLists.txt]: yup, your are right!

**Review by olupton — APPROVED** [review]
LGTM now.

**Review by alexsavulescu — APPROVED** [review]

**Pull request merged**